module/zfs/zvol.c

```diff
@@ -1256,9 +1300,24 @@ zvol_remove_minors_impl(const char *name)
 			 * one is currently using this zv
 			 */
 
-			/* If in use, leave alone */
+			/*
+			 * If in use, try to throw everyone off and try again
+			 * later.
+			 */
 			if (zv->zv_open_count > 0 ||
 			    atomic_read(&zv->zv_suspend_ref)) {
+				zv->zv_flags |= ZVOL_REMOVING;
+				t = taskq_dispatch(
+				    zv->zv_objset->os_spa->spa_zvol_taskq,
+				    zvol_remove_minor_task, zv, TQ_SLEEP);
+				if (t == TASKQID_INVALID) {
+					/*
+					 * Couldn't create the task, so we'll
+					 * do it in place once the loop is
+					 * finished.
+					 */
+					list_insert_head(&delay_list, zv);
+				}
 				mutex_exit(&zv->zv_state_lock);
```

Comment:
I don't know if this is actually an issue, but if the `zvol_remove_minor_task()` taskq gets *immediately* run, I believe`zv` could be freed before the mutex_exit:
 ```C
                         t = taskq_dispatch(
				    zv->zv_objset->os_spa->spa_zvol_taskq,
				    zvol_remove_minor_task, zv, TQ_SLEEP);
				if (t == TASKQID_INVALID) {
					/*
					 * Couldn't create the task, so we'll
					 * do it in place once the loop is
					 * finished.
					 */
					list_insert_head(&delay_list, zv);
				}

< zv could be freed at this point ? >

				mutex_exit(&zv->zv_state_lock);
```
Maybe I'm missing something though.

Reply:
I don't think so, because `zvol_remove_minor_task()` starts by taking `zv_state_lock`, which we already have?

Reply:
I see... hmm.. then wouldn't the last ASSERT() here trip then?
```C
zvol_remove_minors_impl(const char *name)
...
              mutex_enter(&zv->zv_state_lock);                                   

...
                            t = taskq_dispatch(                                  
                                zv->zv_objset->os_spa->spa_zvol_taskq,           
                                zvol_remove_minor_task, zv, TQ_SLEEP); 


static void                                                                      
zvol_remove_minor_task(void *arg)                                                
{                                                                                
       zvol_state_t *zv = (zvol_state_t *)arg;                                   
                                                                                 
       ASSERT(!RW_LOCK_HELD(&zvol_state_lock));                                  
       ASSERT(!MUTEX_HELD(&zv->zv_state_lock));
```

Reply:
Oh, I see the confusion maybe. You said "gets immediately run", but if it's dispatched to a taskq then it's always going to be run on another thread, and that thread won't have the lock, so the asserts pass.

Honestly, those asserts are only there because I also use `zvol_remove_minor_task() as a utility function for processing `delay_list` and also for `zvol_remove_minor_impl()` to do the wait inline. I did try to split it into separate "locked" and "unlocked" versions (one in terms of the other), but the need to drop locks before doing the free made it hard to actually name properly. Maybe a comment on `zvol_remove_minor_task()` to say "sometimes a task, sometimes direct?" would help?

Reply:
> but if it's dispatched to a taskq then it's always going to be run on another thread, and that thread won't have the lock, so the asserts pass.

I'm not 100% sure about that.  Here's the scenario I'm thinking of:
```
--- CPU1 ---                                     --- CPU2 ---
zvol_remove_minor_task()
    mutex_enter(&zv->zv_state_lock) //mutex held
    taskq_dispatch(zvol_remove_minor_task, zv)
    < CPU 1 gets interrupt >                     zvol_remove_minor_task(arg) begins
    < CPU 1 still in interrupt >                 zvol_state_t *zv = (zvol_state_t *)arg
    < CPU 1 still in interrupt >                 ASSERT(!MUTEX_HELD(&zv->zv_state_lock))
                                                 |
                                                 |__ mutex is still held here ???
```

Reply:
(edited :arrow_up_small:)

Reply:
Oh I see what you mean. No, aiui `MUTEX_HELD` is "does this thread hold this mutex", not "is this mutex locked at all". So the assert will never trip if `zvol_remove_minor_task()` is called as a task (it hasn't had the opportunity to take that mutex yet), but provides a little safety for the direct calls. At least, this was my understanding and intent. I'm wondering now if I've fatally misunderstood some deep thing?

Reply:
Oh ok, I get it now.  I assumed MUTEX_HELD meant "someone is holding the mutex", a.k.a `mutex_is_locked()` in the Linux kernel.  But you're right, it actually means "does my thread already hold the mutex":
```C
#define       mutex_owner(mp)             (READ_ONCE((mp)->m_owner))             
#define       mutex_owned(mp)             (mutex_owner(mp) == current)           
#define       MUTEX_HELD(mp)              mutex_owned(mp)
```
Sorry for the confusion!